File: tests/java/CommentsStrings.java
package com.example.comments;

// class FakeInComment { void fakeMethod() {} }

/**
 * Javadoc with fake code:
 *   interface Hidden { void nope(); }
 *   static final int IGNORE = 99;
 */
public class CommentsStrings {

    public static final String GREETING = "Hello, World!";

    // private int notAField = 42;

    private String template;

    public CommentsStrings() {
        // class Inner {} -- this is in a comment, ignore it
        this.template = "class Fake { static final int X = 1; }";
    }

    public String buildQuery(String table) {
        /* interface InBlock { void m(); } */
        String query = "SELECT * FROM " + table + " WHERE class = 'active'";
        return query;
    }

    public String getTextBlock() {
        String block = """
                class InsideTextBlock {
                    static final int FAKE = 100;
                    void fakeMethod() {}
                }
                """;
        return block;
    }

    public int compute(int x) {
        String s = "enum Direction { NORTH, SOUTH }";
        /* enum Hidden { A, B } */
        // record FakeRecord(int a) {}
        return x * 2;
    }

    /*
     * block comment with:
     * public class MultiLineHidden {
     *     public void hiddenMethod() {}
     * }
     */
    public void realMethod() {
        String fake = "public record Ghost(int id) {}";
        System.out.println(fake);
    }
}

/* interface AlsoHidden { void x(); } */

File: tests/scala/comments_strings.scala
package com.example.ignored

// This file tests that declarations inside comments and strings are ignored.

/**
 * class FakeInComment {
 *   def shouldBeIgnored(): Int = 42
 * }
 * trait AlsoIgnored extends FakeInComment
 */
trait RealTrait {
  def realMethod(): String
}

// def notAMethod(): Boolean = true
// val notAVal: Int = 0

object StringLiterals {
  val greeting: String = "def notReal(): Int = 0"
  val template: String = "class Fake { val x = 1 }"

  // Interpolated strings with fake declarations
  val name: String = "world"
  val msg: String = s"trait FakeInInterp { def nope(): Unit }"

  val multiLineIgnored: String = """
    object InTripleQuote {
      val hidden = 42
      def alsoHidden(): Boolean = false
    }
    case class NotReal(x: Int)
  """

  def realDef(x: Int): Int = x + 1
}

class BlockCommentHolder {
  /*
   * object InsideBlockComment {
   *   def hiddenMethod(): String = "nope"
   * }
   */
  val realVal: Double = 3.14

  /* trait FakeBlockTrait { def fake(): Unit } */

  def realMethod(): Boolean = true

  val anotherReal: String = {
    // val fakeInsideBlock: Int = 99
    "result"
  }
}

object NestedStringEdgeCases {
  val tricky: String = "val x = \"nested quote\""
  val raw: String = """class Raw { def m(): Unit = () }"""

  // s"object FakeInLineComment { }"
  val legitimate: Int = 42

  def compute(s: String): Int = s.length
}
